service/storage_service.cc

```diff
@@ -2158,102 +2162,191 @@ future<> storage_service::decommission() {
+                            slogger.warn("removenode[{}]: Failed to send heartbeat", uuid);
+                        }).get();
+                        int nr_seconds = 10;
+                        while (!(*heartbeat_updater_done) && nr_seconds--) {
+                            sleep_abortable(std::chrono::seconds(1), ss._abort_source).get();
+                        }
+                    }
+                    slogger.info("removenode[{}]: Stopped heartbeat_updater", uuid);
+                });
 
-            if (ss._gossiper.get_live_members().contains(endpoint)) {
-                throw std::runtime_error(format("Node {} is alive and owns this ID. Use decommission command to remove it from the ring", endpoint));
-            }
+                // Step 4: Start to sync data
+                req.cmd = node_ops_cmd::removenode_sync_data;
+                parallel_for_each(nodes, [&ss, &req, uuid] (const gms::inet_address& node) {
+                    return ss._messaging.local().send_node_ops_cmd(netw::msg_addr(node), req).then([uuid, node] (node_ops_cmd_response resp) {
+                        slogger.info("removenode[{}]: Got sync_data response from node={}", uuid, node);
+                        return make_ready_future<>();
+                    });
+                }).get();
 
-            // A leaving endpoint that is dead is already being removed.
-            if (tmptr->is_leaving(endpoint)) {
-                slogger.warn("Node {} is already being removed, continuing removal anyway", endpoint);
-            }
 
-            if (!ss._replicating_nodes.empty()) {
-                throw std::runtime_error("This node is already processing a removal. Wait for it to complete, or use 'removenode force' if this has failed.");
+                // Step 5: Announce the node has left
+                std::unordered_set<token> tmp(tokens.begin(), tokens.end());
+                ss.excise(std::move(tmp), endpoint);
+                ss._gossiper.advertise_token_removed(endpoint, host_id).get();
+
+                // Step 6: Finish
+                req.cmd = node_ops_cmd::removenode_done;
+                parallel_for_each(nodes, [&ss, &req, uuid] (const gms::inet_address& node) {
+                    return ss._messaging.local().send_node_ops_cmd(netw::msg_addr(node), req).then([uuid, node] (node_ops_cmd_response resp) {
+                        slogger.info("removenode[{}]: Got done response from node={}", uuid, node);
+                        return make_ready_future<>();
+                    });
+                }).get();
+                *heartbeat_updater_done = true;
+                heartbeat_updater.get();
+                slogger.info("removenode[{}]: Finished removenode operation, removing node={}, sync_nodes={}, ingore_nodes={}", uuid, endpoint, nodes, ignore_nodes);
+            } catch (...) {
+                *heartbeat_updater_done = true;
+                heartbeat_updater.get();
```

Comment:
A safer way to do this is to put these two lines in a defer() block as soon as you create the updater.

Reply:
but defer does not wait for the fiber in the callback to finish

Reply:
Why not?

```c++
auto stop_heartbeat = defer([&] {
    *heartbeat_updater_done = true;
    heartbeat_updater.get();
};
```

Reply:
I was think this:

class deferred_action {
~deferred_action() { if (!_cancelled) { _func(); }; }
}

But the whole thing is inside a thread, so it is ok to call .get() inside the defer.